### PR TITLE
rustdoc: Always warn when linking from public to private items

### PR DESCRIPTION
**Pull request by dennis-hamester**
Change the logic such that linking from a public to a private item always triggers `intra_doc_link_resolution_failure`.
Previously, the warning was not emitted when `--document-private-items` is passed.

This came up during the discussion in https://github.com/rust-lang/rust/pull/74147#discussion_r452597901.

### CODE REVIEW

File: src/test/rustdoc-ui/issue-74134.rs [hunk 1]
```diff
@@ -26,12 +26,14 @@ pub struct Public {
...
 // The following cases are identical to the ones above, except that they are in a private
-// module. Thus they all fall into cases 3 and 4 and should not produce a warning.
+// module. They fall either into the cases 1/2, when --document-private-items is passed, or 3/4
+// when it is not passed.
 
 mod private {
     pub struct Public {
         /// [`super::PublicType`]
         /// [`super::PrivateType`]
+        //[private]~^ WARNING public documentation for `public_item` links to a private item
```
Comment (jyn514) [inline src/test/rustdoc-ui/issue-74134.rs:36]: Hold on, this doesn't look right. This item only shows up when you pass `--document-private-items`, so it shouldn't count as public. I'm not sure if there's an easy way to tell if this is exported though.

**Review by jyn514 — COMMENTED** [review]
Maybe this is a little over-protective after all, the warnings will still show up if you don't pass `--document-private-items`. I'm willing to be convinced either way.

File: src/test/rustdoc-ui/issue-74134.rs (hunk 1 above)
Comment (dennis-hamester) [reply src/test/rustdoc-ui/issue-74134.rs]: Tbh I stumbled over this a couple of times as well, but somehow convinced myself that it is correct this way :roll_eyes:. But now that you mention it, the item is obviously always private, regardless of `--document-private-items`.

I pushed a fixup, which changes the logic again, this time to not rely on the `item`'s visibility anymore. That removes this particular warning and keeps passing all other tests as well.

**Comment by dennis-hamester** [comment]
> Maybe this is a little over-protective after all, the warnings will still show up if you don't pass `--document-private-items`. I'm willing to be convinced either way.

I don't have a strong opinion on this as well. Your comment in https://github.com/rust-lang/rust/pull/74147#discussion_r452597901 did make sense to me and still does imo.

How does the current stabilization plan relate to this? Is it possible to change when a warning triggers after intro-doc links hit stable?

**Comment by jyn514** [comment]
I think my concern is that the lint name `intra_doc_link_resolution_failure` makes it look like the link is broken. But with `--document-private-items` it's actually just a regular warning. Maybe we could distinguish between the two by adding something like this?

```
# for private docs
note: this link will currently works because you passed `--document-private-items`, but will break without
# for public docs
note: this link will resolve properly if you pass `--document-private items`
```

> How does the current stabilization plan relate to this? Is it possible to change when a warning triggers after intro-doc links hit stable?

See https://github.com/rust-lang/rust/issues/43466#issuecomment-643899661:

> When it comes to lints the general Rust stabilty model is that new warn-by-default lints are always okay, deny-by-default lints should usually go through some cycles of warn-by-default unless there are special circumstances.

**Comment by jyn514** [comment]
Another alternative is make this a separate lint altogether: `warn(intra_doc_link_footguns)`. But then the lint would be different with and without private items and I'm not sure if that's great either. Sorry to be so wishy washy :laughing:

**Comment by dennis-hamester** [comment]
I see your point about the lint's name not quite fitting the `--document-private-items` case. But adding an additional lint just for that "fringe" case seems a bit overkill to me, especially when we already have a lint triggering without `--document-private-items`. That flag would then only replace one lint with another.

Considering that, the suggested additional `note:` lines sound like a good middle ground to me. Letting the user know that `--document-private-items` has an influence here is a good UX improvement imo.

There is of course the third option of doing nothing for now.

@jyn514? Maybe @Manishearth has an opinion as well?

**Comment by Manishearth** [comment]
I think an additional note would work! I think it's fine to consider this a "resolution failure".

**Comment by Manishearth** [comment]
Note that when you attempt to import something private from a different crate, rustc throws a similar resolution error with a note iirc.

**Comment by dennis-hamester** [comment]
I went ahead and added notes for both cases as suggested.

**Comment by dennis-hamester** [comment]
Rebased onto `master` and I also took the opportunity and squashed that one fixup commit.

**Pull request merged**